Document to support microsegmentation of VMs using kubeovn subnet ACLs (#843)
harvester/docs: docs/networking/kubeovn-subnet-acls.md

---
sidebar_position: 7
sidebar_label: Kube-OVN Subnet ACLs
title: "Kube-OVN Subnet ACLs"
keywords:
- Harvester
- networking
- Kube-OVN
- access control list
- subnet ACL
---

Isolation between virtual machines is typically achieved using either VLANs (in traditional networks) or virtual switches (in Kube-OVN). If you want to isolate virtual machines within the same virtual switch network, you can enable access control lists (ACLs) on subnets to achieve the required micro-segmentation.

> For more information on using Kube-OVN Subnet ACL and its schema, refer https://kubeovn.github.io/docs/v1.13.x/en/guide/subnet/#subnet-acl https://kubeovn.github.io/docs/v1.13.x/en/reference/kube-ovn-api/#acl

Reviewer: ```suggestion
For more information on using Kube-OVN Subnet ACL and its schema, refer to its [usage](https://kubeovn.github.io/docs/v1.13.x/en/guide/subnet/#subnet-acl) and [API reference](https://kubeovn.github.io/docs/v1.13.x/en/reference/kube-ovn-api/#acl) documentation.
```

## Examples

- Example 1: All virtual machines within the `172.20.10.0/24` subnet, except those with the addresses `172.20.10.1`,`172.20.10.2` and `172.20.10.3` in the subnet range `172.20.10.0/30`, are allowed to communicate with each other.

  ```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.0/30
      priority: 1005
    - action: drop
      direction: from-lport
      match: ip4.src == 172.20.10.0/30
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

> - Example 2: All virtual machines within the `172.20.10.0/24` subnet, except those with the address `172.20.10.1`,`172.20.10.3`, are allowed to communicate with each other. Virtual machines with the address `172.20.10.2` are allowed to communicate because ACL rule execution is based on priority. For this subnet, rules with a priority value of `1006` are executed before `1005`.

Reviewer: ```suggestion
- Example 2: All virtual machines within the `172.20.10.0/24` subnet, except those with the address `172.20.10.1`,`172.20.10.3`, are allowed to communicate with each other. Virtual machines with the address `172.20.10.2` are allowed to communicate because the ACL rules with priority value `1006` takes precedence over that with priority `1005`.
```

```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: allow
      direction: to-lport
      match: ip4.dst == 172.20.10.2
      priority: 1006
    - action: allow
      direction: from-lport
      match: ip4.src == 172.20.10.2
      priority: 1006
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.0/30
      priority: 1005
    - action: drop
      direction: from-lport
      match: ip4.src == 172.20.10.0/30
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

> - Example 3: Virtual machines with the address `172.20.10.2` are allowed to communicate with other virtual machines. However, traffic in the opposite direction is blocked. No virtual machines are allowed to communicate with `172.20.10.2`.

Reviewer: ```suggestion
- Example 3: Virtual machines with the address `172.20.10.2` are allowed to communicate with other virtual machines. However, no virtual machines are allowed to communicate with `172.20.10.2`.
```

---
sidebar_position: 3
sidebar_label: Kube-OVN Subnet ACLs
title: "Kube-OVN Subnet ACLs"
keywords:
- Harvester
- networking
- Kube-OVN
- access control list
- subnet ACL
---

Isolation between virtual machines is typically achieved using either VLANs (in traditional networks) or virtual switches (in Kube-OVN). If you want to isolate virtual machines within the same virtual switch network, you can enable access control lists (ACLs) on subnets to achieve the required micro-segmentation.

Subnet-specific ingress and egress rules (for allowing or dropping traffic) are specified in the `.spec.acls` field of the `Subnet` object.

- match 
  - type: string
    values: source mac address ,destination mac address,source ip address,destination ip address,ip protocol,application port numbers
- action
  - type: string
    value: allow-stateless,allow-related,allow,drop,reject
- direction
  - type: string
    value: from-lport,to-lport
- priority
  - type: integer
     value: 0 to 32767

> ## Features 

Author: Yes right, https://kubeovn.github.io/docs/v1.12.x/en/guide/subnet/#subnet-acl? has internal references to priority and how match conditions are used. Yes we could use it.

- Subnet ACLs consist of an array of rules within the subnet.

- The most important components of the `match` expression are comparisons between symbols and constants (for example, `ip4.dst = 192.168.0.1`, `ip.proto = 6`, `arp.op = 1`, `eth.type = 0x800`, `eth.src = 0a:8f:0a:ec:01:7c`, and `tcp.dst = 9501`).
  
  You can use the logical AND operator `&&` and the logical OR operator `||` to combine comparisons into a larger expression. For more information, see the [Open vSwitch Manual](https://www.ovn.org/support/dist-docs/ovn-sb.5.html).

- ACL rules are evaluated and enforced based on their priority. The higher the priority number, the more precedence the rule takes.

- You can update ACLs on subnets that are used by running virtual machines.

- Deleting ACLs involves removing the contents of the `.spec.acls` field.

## Examples

> - Example 1: Traffic from virtual machines with the IP address `172.20.10.0/30` is blocked for the `vswitch1` subnet. All virtual machines within the `172.20.10.0/24` subnet, except those with the addresses `172.20.10.2` and `172.20.10.3`, are allowed to communicate with each other.

Reviewer: I think the example should focus on the impact the ACL has on the overall traffic flow. Might be worth mentioning the effect it has on the GW IP, since it's also within the 172.20.10.0/30 range.

  ```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.0/30
      priority: 1005
    - action: drop
      direction: from-lport
      match: ip4.src == 172.20.10.0/30
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

> - Example 2: All virtual machines within the `172.20.10.0/24` subnet, except those with the address `172.20.10.3`, are allowed to communicate with each other. Virtual machines with the address `172.20.10.2` are allowed to communicate because ACL rule execution is based on priority. For this subnet, rules with a priority value of `1006` are executed before `1005`.

Author: Rule 1 with priority 1006 allow 172.20.10.2
Rule 2 with priority 1005 blocks 172.20.10.1,17.20.10.2,172.20.10.3

Since Rule 1 is higher priority 172.20.10.2 traffic is still allowed while 172.20.10.1 and 172.20.10.3 are rejected.
Since again this example emphasizes isolation between the VMs, it had no mention about 172.20.10.1

      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

- Example 3: Virtual machines with the address `172.20.10.2` are allowed to communicate with other virtual machines. However, traffic in the opposite direction is blocked. No virtual machines are allowed to communicate with `172.20.10.2`.

  ```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.2
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

- Example 4: TCP traffic originating from port `9501` is blocked on the `vswitch1` subnet (through a combination of rules).

  ```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: drop
      direction: to-lport

>       match: ip4.dst == 172.20.10.2 && tcp.src = 9501

Author: It is an example to show combination of patterns in the same rule.Any inbound and outbound from/to 17.20.10.2 with tcp port 9501 should be blocked.But tcp.src in to-lport is incorrect, it should be updated to tcp.dst = 9501

---

> sidebar_position: 3

Reviewer 2: Please check the `sidebar_position` value of `kubeovn-vpc.md` in [PR 838](https://github.com/harvester/docs/pull/838/files) before modifying this one. Maybe you and Chris can discuss how the Networking sidebar should be organized.

> sidebar_position: 3

Author: Mostly this will come after [PR 838](https://github.com/harvester/docs/pull/838/files), I will confirm with Chris and update

sidebar_label: KubeOVN Subnet ACLs
title: "KubeOVN Subnet ACLs"
keywords:
- Harvester
- KubeOVN
- Subnet ACLs
---

## Micro segmentation between VMs:

Isolation between VMs could be achieved using VLANs traditionally or using virtual switches in KubeOVN.
Enabling  Access Control Lists on subnets further allow micro segmentation of VMs by isolating VMs within the same Virtual Switch Network.

### KubeOVN Subnet ACLs

Edit the subnet spec with acls to apply ingress/egress allow/drop rules per subnet

- match 
  - type: string
    values: source mac address ,destination mac address,source ip address,destination ip address,ip protocol,application port numbers
- action
  - type: string
    value: allow-stateless,allow-related,allow,drop,reject
- direction
  - type: string
    value: from-lport,to-lport
- priority
  - type: integer

>      value: 0 to 32767

Author: @jillian-maroket I think as we have examples below, it is better not to include above under yaml again.I just want users to know about the fields and what values they can specify when they are using subnet acls.WDYT ?

>      value: 0 to 32767

Reviewer 2: This code block appears in the intro section and is preceded by a paragraph that refers to "the `Subnet` object". I'd like users to see the `kind` field for clarity. A few extra lines wouldn't make the doc unnecessarily long.

>      value: 0 to 32767

Author: may be we can refer this "https://github.com/kubeovn/kube-ovn/blob/release-1.13/charts/kube-ovn/templates/kube-ovn-crd.yaml" since we are using kubeovn 1.13 for this release.But again this is a static reference, we need to change again if something changes in upstream.

>      value: 0 to 32767

Author: yes API reference doc is a good one, we might need to change later to later versions when we update kubeovn in Harvester

>      value: 0 to 32767

Reviewer: IMO, changing a link from our doc to their doc is easier and less time consuming than trying to cross-reference with their docs to identify and apply any changes. Also, I am not sure about linking directly to their CRD YAML file because personally, I find CRD files hard to read. Don't they have a 1.13.x API reference doc?

      priority: 1006
    - action: allow
      direction: from-lport
      match: ip4.dst == 172.20.10.2
      priority: 1006
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.0/30
      priority: 1005
    - action: drop
      direction: from-lport
      match: ip4.src == 172.20.10.0/30
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

- **Example 3:** Block traffic only at one direction. VM with `172.20.10.2` will be able to communicate with other VMs, but other VMs will not be able to communicate with `172.20.10.2`

  ```yaml
  apiVersion: kubeovn.io/v1
  kind: Subnet
  metadata:
    name: vswitch1
  spec:
    acls:
    - action: drop
      direction: to-lport
      match: ip4.dst == 172.20.10.2
      priority: 1005
    cidrBlock: 172.20.10.0/24
    excludeIps:
    - 172.20.10.1
    gateway: 172.20.10.1
    gatewayNode: ""
    natOutgoing: false
    private: false
    protocol: IPv4
    provider: vswitch1.default.ovn
    vpc: vpc1
  ```

> - **Example 4:** Block or allow traffic by combining rules.VMs carrying tcp traffic with source port 9501 will be blocked within the `vswitch1` subnet.

Reviewer 2: ```suggestion
- Example 4: TCP traffic originating from port `9501` is blocked on the `vswitch1` subnet (through a combination of rules).
```
